src/GraphQL/Execution/Nodes/ArrayExecutionNode.cs

```diff
@@ -25,35 +25,54 @@ public ArrayExecutionNode(ExecutionNode parent, IGraphType graphType, Field fiel
         }
 
         /// <summary>
-        /// Returns a <see cref="List{T}"/> containing the results of the child execution nodes.
+        /// Returns an object array containing the results of the child execution nodes.
+        /// <see cref="ClearErrorNodes"/> must be called prior to calling this method.
         /// </summary>
         public override object ToValue()
         {
             if (Items == null)
                 return null;
 
             var items = new object[Items.Count];
+            for (int i = 0; i < Items.Count; ++i)
+            {
+                items[i] = Items[i].ToValue();
+            }
+
+            return items;
+        }
+
+        /// <inheritdoc/>
+        public override bool ClearErrorNodes()
+        {
+            if (Items == null)
+                return true;
+
+            if (Items.Count == 0)
+                return false;
+
+            var isNullableType = false;
+
             for (int i = 0; i < Items.Count; ++i)
             {
                 var item = Items[i];
-                object value = item.ToValue();
+                bool valueIsNull = item.ClearErrorNodes();
 
-                if (value == null)
+                if (valueIsNull && !isNullableType)
                 {
-                    var listType = item.FieldDefinition.ResolvedType;
-                    if (listType is NonNullGraphType nonNull)
-                        listType = nonNull.ResolvedType;
-
-                    if (((ListGraphType)listType).ResolvedType is NonNullGraphType)
+                    if (((ListGraphType)GraphType).ResolvedType is NonNullGraphType)
```

Comment:
Strange. It seems that evaluation of this expression can be taken out of the loop : 
```
    var isNullableType = ((ListGraphType)GraphType).ResolvedType is NonNullGraphType;
```

Reply:
Yes it can, but the loop still needs to executed. I benchmarked it a few different ways and found this to be the fastest. Of course it would really depend on what exact data was being returned.

Reply:
Strange. Why the fastest? Maybe only for arrays for size=0? :) But in general, I doubt that these checks are necessary. It seems to me that they are already executed earlier at the execution stage. Or at least we need to add them there.

Reply:
It is the fastest because often there are no situations where the value is null, and the arrays are often small.  So a couple quick null checks ends up being faster than a (relatively) costly type check -- followed by the same null check.  At least that's my theory.

I don't know how it can determine it earlier in the execution.  The child nodes have not been executed at that point in time.  Child nodes may have non-null "grandchildren" which return null, causing a cascade up to the "child".  That only happens after the nodes have been executed.  Sure if the children are value execution nodes then it could be checked at the time it is set, but it's got to be checked at some point in time.

The only thing that could happen is that when a value execution node is set, it could check all the parent nodes at the time to clear the children.  But this may prevent some child nodes from executing.  So you might need to add a boolean property.  It just seems like we're going in a loop here.  If you want to try it, there may be some way to get it to run a tad faster.

Reply:
> But this may prevent some child nodes from executing.

Is it bad? I think this is a win.

Reply:
> > But this may prevent some child nodes from executing.
> 
> Is it bad? I think this is a win.

Your suggestion meets spec.

> If during `ExecuteSelectionSet()` a field with a non‐null fieldType throws a field error then that error must propagate to this entire selection set, either resolving to null if allowed or further propagated to a parent field.
>
> If this occurs, any sibling fields which have not yet executed or have not yet yielded a value may be cancelled to avoid unnecessary work.

Reply:
OK. #2253